ESQL: Fix for overzealous validation in case of invalid mapped fields

alex-spies opened the pull request:
Fix https://github.com/elastic/elasticsearch/issues/111452

https://github.com/elastic/elasticsearch/pull/107545 postpones some of the validation for fields that occur in multiple indices with different types. Such fields must not be explicitly used in the query.

The validation now happens [after the main reference resolution](https://github.com/elastic/elasticsearch/pull/107545/files#diff-2c689d9c0606d310fd36a603064f43cb4b23d356f08983fbdb98c65e76e6919dR1187-R1192). This is problematic for renames in particular. Consider a field `a` that has the same type in all indices:
```
FROM idx | RENAME a AS b
```
is perfectly valid. But after reference resolution, the `RENAME` becomes a `PROJECT` that has to list _all_ fields that it will output. If there is a `field` that is `keyword` in `index1` but `ip` in `index2`, we have the following
```
FROM idx | PROJECT field, a AS b
```
Performing the validation _now_ gives wrong results, because it sees `field` and flags it as invalid; `field` was not used explicitly, but ended up in the plan due to replacing `RENAME` by a `PROJECT`.

~~This PR reverts to the original approach: during the initial resolution in `ResolveRefs`, we immediately validate usage of a field attribute referring to a multi-typed field. Usage is considered valid if the attribute is either used inside a conversion function, like `TO_IP`, or if the attribute is being dropped.~~

**Updated with another approach**: This PR now aligns how we treat multi-typed fields with how we treat unsupported field types. The latter are already being passed through and are even allowed to be mentioned explicitly in `KEEP`, ~~`RENAME`~~ and `DROP`. To achieve this, rather than mapping `FieldAttribute`s containing an `InvalidMappedField` to `UnresolvedAttribute` - which the verifier finds and counts as invalid query - we map such field attributes to `UnsupportedAttribute`, which is what fields with unsupported types are also being mapped to.

To avoid weird workarounds when resolving `STATS`, type resolution for convert functions becomes more lenient, so that multi-typed field attributes (`FieldAttribute` containing an `InvalidMappedField`) are considered resolved as long as all of the mapped types are supported by the convert function; e.g. `TO_IP(client_ip)` in case `client_ip` is mapped as both `keyword` and `ip`.

alex-spies committed:
Quick fix: make validation more lenient

alex-spies committed:
Fix csv test

alex-spies committed:
Disallow union types EXCEPT when they're cast

This is more conceptually correct: union typed attributes must not be
used in a query, unless they're immediately cast to a single type or
dropped.

Implementation is haaacky and needs improvement, but it gets the job
done.

alex-spies commented:
The current version now reverts back to how we validated multi-typed fields before we added union types:
- While resolving refs, we try to map unresolved attributes to field attributes.
- If we find one, but it contains an `InvalidMappedField`, we'll generally not resolve this, causing a validation error if someone explicitly uses multi-typed fields,
- unless the multi-typed field is inside a conversion function like `to_string`, or if it's inside a `DROP`.

This should be correct from the get-go because it performs the resolution and validation directly on the _parsed_ logical plan, rather than after the plan has already been altered by resolution steps.

alex-spies committed:
Add capability for the fix

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -874,8 +904,19 @@ private static List<Attribute> potentialCandidatesIfNoMatchesFound(
         return matches;
     }
 
-    private static Attribute handleSpecialFields(UnresolvedAttribute u, Attribute named) {
-        return named.withLocation(u.source());
+    private static Attribute tryResolveUnionTypeFields(UnresolvedAttribute toResolve, Expression parent, Attribute resolveTo) {
+        if (resolveTo instanceof FieldAttribute fa) {
+            // incompatible mappings
+            var field = fa.field();
+            // Multi-typed fields are allowed if and only if we convert them to a single type immediately.
+            if (field instanceof InvalidMappedField imf && (parent instanceof AbstractConvertFunction == false)) {
+                resolveTo = toResolve.withUnresolvedMessage(
+                    "Cannot use field [" + fa.name() + "] due to ambiguities being " + imf.errorMessage()
+                );
+            }
+        }
+
+        return resolveTo.withLocation(toResolve.source());
```
`handleSpecialFields` used to look for FieldAttributes mapped with multiple types before https://github.com/elastic/elasticsearch/pull/107545. This brings back this logic, but takes into account the attribute's parent because multi-typed FieldAttributes are allowed if they're immediately cast.

alex-spies committed:
Fix message on invalid union types for conversion

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

alex-spies committed:
Skip csv test unless run as IT

alex-spies committed:
Add tests for KEEP

alex-spies committed:
More tests

alex-spies committed:
Even more tests

alex-spies committed:
Update docs/changelog/111475.yaml

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Hi @alex-spies, I've updated the changelog YAML for you.

alex-spies committed:
Update docs/changelog/111475.yaml

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java:
```diff
@@ -176,19 +177,35 @@ public static TypeResolution isType(
+            return TypeResolution.TYPE_RESOLVED;
+        }
+
+        if (e instanceof FieldAttribute fa && fa.field() instanceof InvalidMappedField imf) {
+            return new TypeResolution(errorStringIncompatibleTypes(operationName, paramOrd, name(e), imf.errorMessage(), acceptedTypes));
+        }
```
This improves error messages for failed type resolutions in case of union types, see yaml tests.

alex-spies committed:
Update capabilities in tests

alex-spies committed:
Do not use UnresolvedAttribute without a field

alex-spies committed:
Revert Analyzer to main

alex-spies committed:
More tests

alex-spies committed:
Simpler approach

alex-spies committed:
Remove test again

used only for reference

alex-spies committed:
Fix bug in ResolveRefs, update tests

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

alex-spies commented:
I took a new approach today; this should be closer to what we already had, because it uses the pass-through mechanism that we have in place for fields with unsupported data types (passed through as `null`s). It also avoids any changes to `Node`.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1067,24 +1068,19 @@ public static Expression castStringLiteral(Expression from, DataType target) {
                     target,
                     e.getMessage()
                 );
-                return new UnsupportedAttribute(
-                    from.source(),
-                    String.valueOf(from.fold()),
-                    new UnsupportedEsField(String.valueOf(from.fold()), from.dataType().typeName()),
-                    message
-                );
```
Unrelated, but maybe it's better not to fake an ES field here, `UnresolvedAttribute` seems to do the trick as well.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1120,24 +1116,6 @@ private LogicalPlan doRule(LogicalPlan plan) {
                 return plan;
             }
 
-            // In ResolveRefs the aggregates are resolved from the groupings, which might have an unresolved MultiTypeEsField.
-            // Now that we have resolved those, we need to re-resolve the aggregates.
-            if (plan instanceof Aggregate agg) {
-                // TODO once inlinestats supports expressions in groups we'll likely need the same sort of extraction here
-                // If the union-types resolution occurred in a child of the aggregate, we need to check the groupings
-                plan = agg.transformExpressionsOnly(FieldAttribute.class, UnionTypesCleanup::checkUnresolved);
-
-                // Aggregates where the grouping key comes from a union-type field need to be resolved against the grouping key
-                Map<Attribute, Expression> resolved = new HashMap<>();
-                for (Expression e : agg.groupings()) {
-                    Attribute attr = Expressions.attribute(e);
-                    if (attr != null && attr.resolved()) {
-                        resolved.put(attr, e);
-                    }
-                }
-                plan = plan.transformExpressionsOnly(UnresolvedAttribute.class, ua -> resolveAttribute(ua, resolved));
-            }
```
This may have been already obsolete; in any case, there shouldn't be a need to special-case aggs while resolving union types.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -1264,7 +1230,14 @@ public LogicalPlan apply(LogicalPlan plan) {
+                    new UnsupportedEsField(imf.getName(), types),
+                    unresolvedMessage,
+                    fa.id()
+                );
```
This is the crux of the fix.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -461,7 +460,9 @@ private LogicalPlan resolveStats(Stats stats, List<Attribute> childrenOutput) {
                 for (Expression e : groupings) {
-                    Attribute attr = Expressions.attribute(e);
+                    // In case of union types, the expression may contain yet-to-be resolved conversion functions, so do not perform a
+                    // type check here.
+                    Attribute attr = Expressions.attributeUnchecked(e);
```
Oh no, forcing a reference attribute here is also wrong, because in case of truly unresolved expressions, calling `.datatype()` is wrong. I'll figure out a way around this.

costin replied:
Thanks for the explanation - replace the Alias logic here; if it's an alias , return a ReferenceAttribute otherwise call toAttribute().
This gets rid of the attributeUnchecked() method

costin reviewed (changes requested):
Thanks for moving this forward Alex - great stuff.
Don't generalize toAttributeUnchecked - rather perform the Alias over unresolved into the stats resolution.
I wonder though if this can occur in other places where we have named union types..

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/NamedExpression.java:
```diff
@@ -55,8 +55,18 @@ public boolean synthetic() {
+     * Return either {@code this} if it is an {@link Attribute}, or a {@link ReferenceAttribute} to it otherwise.
+     * Do not return an {@link UnresolvedAttribute} if this is unresolved.
+     */
+    public abstract Attribute toAttributeUnchecked();
```
Yeah, I think this needs to go; I think I found a better way: be more precise in how we perform the resolution of the aggregates in `resolveStats` based on the already resolved groupings.

alex-spies replied:
Nevermind; I had to go another route after all. I _wanted_ to only resolve the aggregates if the groups could be resolved - but that created wrong error messages, as resolvable aggregates were treated as unresolvable.

alex-spies committed:
Fix resolveStats

alex-spies committed:
One more test

alex-spies committed:
Remove toAttributeUnchecked

alex-spies committed:
Another approach: make TO_IP(client_ip) resolved

Make the type resolution more lenient for AbstractConvertFunction; we
know TO_IP(client_ip) is effectively resolved if client_ip is a
FieldAttribute containing an InvalidMappedField iff the
InvalidMappedField contains only types that are accepted by TO_IP.

This side steps lots of tricky issues during resolution of stats.

alex-spies commented:
One more approach that seems to solve the problems when resolving aggregates:
Consider expressions like `TO_IP(client_ip)` _resolved_ when `client_ip` is a multi-typed field attribute whose types can all be used in `TO_IP`. Considering it unresolved caused issues when trying to resolve `STATS` and `INLINESTATS`, which would've been hacky to work around; considering `TO_IP(client_ip)` resolved is conceptually cleaner, as we _know_ this is valid.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

alex-spies committed:
Make test skip more lenient

alex-spies committed:
Update docs

astefan reviewed (commented):
I think it's ok to allow these unsupported fields in `keep` and `rename` (`drop` was already allowed) but we need to be careful to the conceptual side effects (I don't know if there are any yet) meaning we must have a good story around allowing unsupported data types in these additional commands. So far, we said "this is unsupported, so I want to remove it, `drop` gives me that", so what's the argument for `keep` and especially `rename`?

I've tested various things:

1.
```
        "properties": {
            "x": {
                "type": "text"
            },
            "flatx": {
                "type": "flattened"
            } 
        }
```
`from test | rename flatx as whatever`

results in 
```
"Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null"
```
which is **the same behavior in `main` and in this PR**.

2. `FROM sample_data, sample_data_str | rename client_ip as foo | sort bla`

results in

```
path: /_query, params: {format=txt, error_trace=true}, status: 500 org.elasticsearch.xpack.esql.core.capabilities.UnresolvedException: Invalid call to dataType on an unresolved object ?bla
        at org.elasticsearch.xpack.esql.core.expression.UnresolvedAttribute.dataType(UnresolvedAttribute.java:101)
        at org.elasticsearch.xpack.esql.core.expression.Attribute.canonicalize(Attribute.java:146)
        at org.elasticsearch.xpack.esql.core.expression.Expression.canonical(Expression.java:161)
        at org.elasticsearch.xpack.esql.core.expression.Expressions.canonicalize(Expressions.java:96)
        at org.elasticsearch.xpack.esql.core.expression.Expression.canonicalize(Expression.java:170)
        at org.elasticsearch.xpack.esql.core.expression.Expression.canonical(Expression.java:161)
        at org.elasticsearch.xpack.esql.core.expression.ExpressionSet.add(ExpressionSet.java:81)
        at org.elasticsearch.xpack.esql.optimizer.rules.PruneRedundantSortClauses.rule(PruneRedundantSortClauses.java:24)
        at org.elasticsearch.xpack.esql.optimizer.rules.PruneRedundantSortClauses.rule(PruneRedundantSortClauses.java:17)
```

which is **the same behavior in `main` and in this PR**.

3. `FROM sample_data, sample_data_str | rename client_ip as foo` which is more relevant to this PR:

**in main** `"Cannot use field [client_ip] due to ambiguities being mapped as [2] incompatible types: [ip] in [sample_data], [keyword] in [sample_data_str]"`

**in PR** 

```
"Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null"
```

alex-spies committed:
Minor refactor

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

alex-spies committed:
Add test

alex-spies commented:
`FROM sample_data, sample_data_str | rename client_ip as foo` excellent find @astefan (How could I miss that..).

The error message that this yields

```
"Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null"
```

is consistent with what you get when you rename a field of unsupported type; in your example, that'd be
```
from test | rename flatx as foo
```

I'll try to improve that; renaming should be okay for both or have a nice error message for both. I'm leaning towards error message, because otherwise users might wanna do `FROM idx | RENAME multi_typed_field AS foo | EVAL foo = to_ip(foo)`.

The argument for allowing multi-typed field in more places, i.e. `KEEP`, is that we should not be _more_ restrictive with multi-typed fields than with fields of unsupported types. Mapping `FieldAttribute`s containing `InvalidMappedField`s (i.e. union type fields) to `UnsupportedAttribute` forces consistency for these two kinds of fields.

In particular, `KEEP flatx` is allowed in your example. And allowing this is _important_ because on main, `KEEP *x` _fails_ which is a bad UX imho - I wanna be able to `KEEP` a bunch of fields matching a wildcards; if this accidentally includes a couple multi-typed fields, I as a user won't care so much; but having to explicitly `DROP` them before the `KEEP` is inconvenient, and it's inconsistent because it's not needed for fields of unsupported type, like `flatx`.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -458,17 +456,17 @@ private LogicalPlan resolveStats(Stats stats, List<Attribute> childrenOutput) {
             }
 
-            if (stats.expressionsResolved() == false) {
-                AttributeMap<Expression> resolved = new AttributeMap<>();
```
The role of AttributeMap is to guarantee equality - while it can deal with ordering (it's a linkedmap), it's better to have this recorded elsewhere such as a list.

alex-spies replied:
Ah, fair point; however, there shouldn't be groupings with same name id but different names, so the `mergeOutputAttributes` below should take care of deduplicating fine.

I think using the `AttributeMap` here is less obvious (I completely forgot the thing preserves iteration order), esp. since an `AttributeSet` seems to suffice.

costin commented:
> So far, we said "this is unsupported, so I want to remove it, drop gives me that", so what's the argument for keep and especially rename?

Unsupported fields are allowed to pass through. With keep we allow for the them to be retained, with drop to be discarded and with rename to change their alias. All valid sub-cases of projection.
However unsupported fields cannot be manipulated.

> FROM idx | RENAME multi_typed_field AS foo | EVAL foo = to_ip(foo)

Can the query be written differently to not throw an error? Is the issue EVAL or RENAME?

alex-spies commented:
> with drop to be discarded and with rename to change their alias. All valid sub-cases of projection.
However unsupported fields cannot be manipulated.

That's all true except for `RENAME`; this currently does not work with unsupported types in main (layout errors). We can fix it or forbid it. In any case, I'd like to do the same for multi-typed fields and fields of unsupported type for consistency.

> `FROM idx | RENAME multi_typed_field AS foo | EVAL foo = to_ip(foo)`
> Can the query be written differently to not throw an error? Is the issue EVAL or RENAME?

This can be simply written as `FROM idx | EVAL foo = to_ip(multi_typed_field)`. As of now, the problem is the layout issue with RENAME used on unsupported attributes; but even if fixed, it'd be a bit of additional work to ensure renaming a multi-typed field _and then_ using a conversion function on it is properly supported - at the very least we'd need to add good tests.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

alex-spies committed:
D'oh

alex-spies committed:
Fix RENAME for unsupported/multi-typed fields

alex-spies commented:
Found a fix for `RENAME unsupported AS foo` which also fixes `RENAME multi_typed_field AS foo`. It required creating reference attributes from aliases even if the child expression is not fully resolved - which is exactly the case for `UnsupportedAttribute`, which is considered unresolved, but its type is resolved (to `UNSUPPORTED`).

@costin @astefan let's discuss whether we want to allow `RENAME` for unsupported/multi types before moving on with this fix.

(Side note: maybe the verification step could be simplified if we didn't consider `UnsupportedAttribute` as unresolvable. I suspect that the mechanism that allows passing through unsupported types could be simpler if we shifted the responsibility to the type resolution step only, i.e. don't fail a query just because we see an `UnsupportedAttribute`, but fail it when this is used in a place that requires something other than the `UNSUPPORTED` data type. It would prevent having to change `Alias.toAttribute` as well.)

alex-spies committed:
Continue validation after RENAME multityped AS foo

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/local/EsqlProject.java:
```diff
@@ -65,8 +66,12 @@ public EsqlProject replaceChild(LogicalPlan newChild) {
         for (NamedExpression projection : projections()) {
-            // don't call dataType() - it will fail on UnresolvedAttribute
-            if (projection.resolved() == false && projection instanceof UnsupportedAttribute == false) {
+            // Pass through for unsupported types.
+            if (projection instanceof UnsupportedAttribute
+                || (projection instanceof Alias a && a.child() instanceof UnsupportedAttribute)) {
+                continue;
+            }
+            if (projection.resolved() == false) {
```
@astefan this fixes validation for the query `FROM sample_data, sample_data_str | rename client_ip as foo | sort bla` (very good find!).

If we're happy with the current approach, I'll add a verifier test for it.

alex-spies committed:
Address Andrei's remarks

alex-spies commented:
@astefan found another nice edge case:

```
// client_ip is mapped as ip and keyword
FROM sample_data, sample_data_str | rename client_ip as foo | sort foo
```
In the current state of the PR, this actually computes, while
```
FROM sample_data, sample_data_str | sort client_ip
```
Correctly returns a validation error. The same is true when we replace the `sort` by an `mv_expand`.

This indicates that allowing to `RENAME my_unsupported_field AS foo` will require more work because we'll at least have to carefully re-check the validation and add additional tests.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

alex-spies committed:
Invalidate UnsupportedAttribute in RENAME

alex-spies committed:
Add verifier test

alex-spies committed:
Remove csv test

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

alex-spies committed:
Checkstyle

alex-spies committed:
Use lower cased type name in ENRICH validation msg

alex-spies committed:
Remove obsolete tests

alex-spies committed:
D'oh

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

astefan commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec:
```diff
@@ -1015,3 +1015,256 @@ client_ip:ip | event_duration:long |    message:keyword    |    @timestamp:keywo
+required_capability: union_types_fix_rename_resolution
+FROM sample_data, sample_data_ts_long
+// This is not allowed because it implicitly uses @timestamp, which is a multi-typed field.
+// It probably should be! Otherwise we're inconsistent with `KEEP *`.
```
Is this comment still applicable?

alex-spies replied:
Good catch, thank you!

astefan reviewed (approved):
LGTM 👍

alex-spies committed:
Remove leftover

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-union-types-breaking-unsupported-in-renames

costin reviewed (approved):
👍

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -2018,6 +2018,19 @@ M
 null
 ;
 
+shadowingInternalWithGroup2#[skip:-8.14.1,reason:implemented in 8.14]
```
I guess this is not related, but good to add a shadowing test within the groups in particular.

alex-spies replied:
This could've tripped due to the logic in the resolution of STATS, which _didn't look_ like it correctly tracks the order of the groups. Turns out it did, because `AttributeMap` uses a `LinkedHashMap` internally; but better to have something to guard against regressions.

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.15  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 111475`

alex-spies commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/111735">](https://github.com/elastic/elasticsearch/pull/111735)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
